Lesson02 loops

dmidob opened the pull request.

dmidob committed:
Package renamed to camelCase
Package with tests added to VCS

dmidob committed:
ArrayGenerating and ArrayTypeConversion classes added

dmidob committed:
Verbs in classes names renamed to nouns

dmidob committed:
ArrayEditorTests class renamed to ArrayEditorParametrizedTests

dmidob committed:
Variables for test arrays (MaxMinAverageTests class) renamed

dmidob committed:
Constnt added in ArrayGenerator class for right bound limit value

dmidob committed:
Variables renamed for PrimeNumbersTests by "actual" and "expected" postfixes added.
Equals assertion is using 2 List-typed arrays now.

dmidob committed:
Added test class with dataprovider - ArrayEditorDataProviderTests

dmidob committed:
Subpackages added and corresponding classes moved to them.
Code formatted (identions and empty lines were reviewed).

dmidob committed:
Variable renamed

dmidob committed:
Methods renamed

dmidob committed:
lesson 02 tasks resolving

dmidob committed:
lesson 02 tasks resolving - without unit tests

dmidob committed:
Merge branches 'lesson02_loops' and 'master' of https://github.com/dmidob/java_learn into lesson02_loops

 Conflicts:
	src/main/java/lesson01/loopspracticing/PrimeNumbers.java

dmidob committed:
merge conflicts resolving

ArtBi commented on src/main/java/lesson02/CharsCounter.java:
```diff
@@ -0,0 +1,45 @@
+        } else {
+            System.out.println("Char '" + charToFind + "' in '" + inputString + "' string is present "
+                    + countOfCharToFind + " time(s)");
+        }
```
It looks like a good candidate to move a separate method.

fanik replied:
Additionally, those strings are good candidates for class constants.
See how you may use them with your insertions:
https://www.geeksforgeeks.org/java-string-format-examples/

dmidob replied:
Class has been rewritten

ArtBi commented on src/main/java/lesson02/ConsoleInput.java:
```diff
@@ -0,0 +1,21 @@
+
+public class ConsoleInput {
+    static int inputNumberFromConsole (String textMessageForInput) {
+        Scanner scanner = new Scanner(System.in);
```
A good point to cache this scanner object. (use the singleton pattern)

fanik replied:
If Singleton is used - which pros and cons it would give to you?
Are there any other alternatives?
P.S. do not spend much time on this tonight, we can discuss it during the meeting.

dmidob replied:
Get some info of how to work with Singleton pattern. Pros and cons.

dmidob committed:
Randomize class added.
New task (randoms printer) added

dmidob committed:
Randomize class added.
New task (randoms printer) added

ArtBi commented on src/main/java/lesson02/EvensRangePrinter.java:
```diff
@@ -0,0 +1,28 @@
+                    System.out.print(i + " ");
+                }
+            }
+        }
```
It looks like a couple of logic levels (Checking and printing)

dmidob replied:
This was fixed. 2 separate methods were added

ArtBi commented on src/main/java/lesson02/PalindromeChecker.java:
```diff
@@ -0,0 +1,31 @@
+        System.out.println(isPalindromeChecking("12345 6 7 6 54321"));
+    }
+
+    private static boolean isPalindromeChecking (String inputString){
```
Required and actual method names aren't the same.

dmidob replied:
Method and boolean variable renamed

fanik commented on src/main/java/lesson02/CharsCounter.java:
```diff
@@ -0,0 +1,45 @@
+
+    //alternative method for chars counter with beautiful output
+    private static void countCharsWithOutputMessage (String inputString, char charToFind){
+        int countOfCharToFind = 0;
```
You can avoid code duplication by reusing `countChars` + extra code afterward. Recall DRY principle.

dmidob replied:
that method was removed

fanik commented on src/main/java/lesson02/ConsoleInput.java:
```diff
@@ -0,0 +1,21 @@
+package lesson02;
+
+import java.util.Scanner;
+
```
When first looking at this class during the review I wasn't able to identify that it is a utility class for some other task, think how this concern can be addressed within class/package naming?

dmidob replied:
Utility classes moved to a new subpackage. Comments also added

fanik commented on src/main/java/lesson02/EvensRangePrinter.java:
```diff
@@ -0,0 +1,28 @@
+public class EvensRangePrinter {
+
+    public static void main (String[] args)  {
+        printEvensRange(ConsoleInput.inputNumberFromConsole("Please input first number: "),
```
My personal preference for arranging long function arguments or many of them is as described here:
https://stackoverflow.com/a/6023487

I haven't found a strict rule for that in java, but you can read the thread to get your opinion on this topic.

dmidob replied:
That was fixed

fanik commented on src/main/java/lesson02/PalindromeChecker.java:
```diff
@@ -0,0 +1,31 @@
+        int i = 0;
+        while (i < inputString.length()) {
+            if (inputString.charAt(i) == inputString.charAt(inputString.length() - i - 1)){
+                isPalindrome = true;
```
can you rewrite code without setting this value?

dmidob replied:
the method was rewritten without additional boolean value

fanik commented on src/main/java/lesson02/PrimeChecker.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class PrimeChecker {
+    public static void main (String[] args) {
+        if (isPrimeNumberChecking(ConsoleInput.inputNumberFromConsole("Please input " +
```
Create a habit, if you see inline text - move it to constant

dmidob replied:
fixed

fanik commented on src/main/java/lesson02/PrimeChecker.java:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    private static boolean isPrimeNumberChecking (int n){
```
Which extra meaning added by `Checking` in the function name?

dmidob replied:
This was used to avoid method and boolean variable naming duplication. Since variable was removed - method renamed

fanik commented on src/main/java/lesson02/RangePrinter.java:
```diff
@@ -0,0 +1,33 @@
+            for (int i = firstNumber; i <= lastNumber ; i++) {
+                System.out.print(i + " ");
+            }
+        }
```
When you have else statement, you can format as follows:
} else {

This doesn't break the styling and being more compact.

dmidob replied:
fixed

fanik commented:
No tests (cry)

dmidob committed:
2 subpackages created.
Tasks and utility classes sorted

dmidob committed:
SumApp and MultiplicationTablePrinter tasks resolve added

dmidob committed:
2 separate methods created for EvensRangePrinter
2 additional classes with task resolving added (missed add files during previous commit)

dmidob committed:
boolean variable renamed (to avoid name duplication with method)

dmidob committed:
CharsCounter, PalindromeChecker and PrimeChecker classes have been rewritten;

dmidob committed:
Some minor formatting ("else" format)

dmidob committed:
PrimeChecker method renamed

dmidob committed:
Typos fixing

ArtBi commented on src/main/java/lesson02/tasksresolve/EvensRangePrinter.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private static void printEvensRange (int firstNumber, int lastNumber){
+        if (isInputEligible(firstNumber, lastNumber)){
```
I suggest renaming isInputEligible to isInputValid or something more familiar.

dmidob replied:
"Eligible" is more appropriate here

ArtBi commented on src/main/java/lesson02/tasksresolve/EvensRangePrinter.java:
```diff
@@ -0,0 +1,46 @@
+        if (num1 < num2){
+            isEligible = true;
+        }
+        return isEligible;
```
You can just return the result of the comparison.

dmidob replied:
This was fixed

ArtBi commented on src/main/java/lesson02/tasksresolve/MultiplicationTablePrinter.java:
```diff
@@ -0,0 +1,25 @@
+
+    static void printMultiplicationTable (int size){
+        for (int i = 1; i < size + 1; i++) {
+            for (int j = 1; j < size + 1; j++) {
```
You can use 'i <= size' instead of 'i < size + 1'

dmidob replied:
Agree. I'm wondering how can I missed this during the coding )

ArtBi commented on src/main/java/lesson02/tasksresolve/PalindromeChecker.java:
```diff
@@ -0,0 +1,30 @@
+        int i = 0;
+        while (i < inputString.length()) {
+            if (inputString.charAt(i) != inputString.charAt(inputString.length() - i - 1)){
+                isPalindrome = false;
```
can you rewrite code without setting this value? (fanik ©)

dmidob replied:
Yes I can (now :) )

ArtBi commented on src/main/java/lesson02/tasksresolve/PrimeChecker.java:
```diff
@@ -0,0 +1,56 @@
+    private static final String MESSAGE_FOR_NUMBER_IS_LESS_THAN_ZERO = "Number should be > 0";
+
+    public static void main (String[] args) {
+        if (isPrime(ConsoleInput.inputNumberFromConsole(MESSAGE_FOR_NUMBER_INPUT))){
```
I suggest moving to the local variable result of the input from the console.

dmidob replied:
Fixed

ArtBi commented on src/main/java/lesson02/tasksresolve/PrimeChecker.java:
```diff
@@ -0,0 +1,56 @@
+        if (numberForPrimeChecking < 0) {
+            isNumberPositive = false;
+        }
+        return isNumberPositive;
```
You can do without the boolean variable. (return result of comparison)

dmidob replied:
Method has been rewritten.

ArtBi commented on src/main/java/lesson02/tasksresolve/RandomsPrinter.java:
```diff
@@ -0,0 +1,31 @@
+            isError = true;
+        }
+        return isError;
+    }
```
You can extract creating one random number to another method. (For more scalability)

dmidob replied:
Actually My code has already such functionality. As discussed during code review - there was a misunderstanding of that :)

ArtBi commented on src/main/java/lesson02/tasksresolve/RandomsPrinter.java:
```diff
@@ -0,0 +1,31 @@
+        printRandomNumbers(3, 0, 20 );
+    }
+
+    static boolean printRandomNumbers (int countOfRandomNumbers, int minNumber, int maxNumber){
```
I suggest changing method signature to min, max, countOfNumbers

dmidob replied:
The task said so to use method signature in that way

ArtBi commented on src/main/java/lesson02/tasksresolve/RangePrinter.java:
```diff
@@ -0,0 +1,42 @@
+                    ConsoleInput.inputNumberFromConsole(MESSAGE_FOR_LAST_NUMBER_INPUT)
+            );
+        } catch (Exception e) {
+            System.exit(-1);
```
More useful print some message in case exception or print stack trace.

dmidob replied:
Fixed

dmidob committed:
PR comments fixing

dmidob committed:
Unit tests added

fanik commented on src/main/java/lesson02/tasksresolve/RandomsPrinter.java:
```diff
@@ -0,0 +1,31 @@
+
+    static boolean printRandomNumbers (int countOfRandomNumbers, int minNumber, int maxNumber){
+        boolean isError = false;
+        if (countOfRandomNumbers > 0 & minNumber <= maxNumber){
```
Are you sure you want to use "&" here?

dmidob replied:
of course no. it has been done in a rush

fanik commented on src/main/java/lesson02/tasksresolve/SumApp.java:
```diff
@@ -0,0 +1,33 @@
+
+    public static int clientElemsSum (){
+        int sumOfInputNumbers = 0;
+        for (int i = 1; i > 0; i++) {
```
"For" - is usually used when you need to iterate over a fixed amount of items.
While - is more sufficient in this case.

dmidob replied:
I used "For" loop because there is a requirement that we need to break the loop once the input number is equal to 0. Since I'm calling number input from the console within the loop I'm not able to use (inputNumber!=0) condition in "while" statement.
